agasc/supplement/utils.py

```diff
@@ -134,3 +144,191 @@ def save_version(filename, table_name):
                    append=True, overwrite=True)
     last_updated.write(str(filename), format='hdf5', path='last_updated',
                        append=True, overwrite=True)
+
+
+def update_table(filename, table, path, dtype, keys=['agasc_id'], dry_run=False, create=False):
+    """
+    Update a table of the AGASC supplement.
+
+    This overwrites all rows already in the supplement table and appends that rows not yet in the
+    supplement table.
+
+    :param filename:
+    :param table: table.Table
+    :param path: str
+        the path of the table in the supplement ('obs', 'mags')
+    :param dtype: np.dtype
+        the dtype of the table. Used only if the table is not in the supplement already.
+        If it is not given, and the table does not exist in the supplement,
+        an exception will be raised.
+    :param keys: list
+        a list of columns that uniquely identify a row in the table.
+        In case of duplicates, the first appearance is kept.
+    :param dry_run: bool
+    :param create: bool
+        Create a supplement file if it does not exist
+    """
+    filename = Path(filename)
+
+    if len(table) == 0:
+        logger.info('Nothing to update')
+        return
+
+    suppl_table = get_table(filename, path, dtype, create=create)
+    table = Table(table if len(table) else None, dtype=dtype)
+    table = unique(table, keys=keys, keep='first')
+
+    # for now I require no masked elements.
+    if table.mask is not None and np.any(table.mask):
+        masked = [name for name in table.colnames if np.any(table[name].mask)]
+        raise Exception('"{path}" table should have no masked elements, '
+                        'but element in these columns are masked: ',
+                        ', '.join(masked))
+
+    # entries already in supplement
+    intersect = (table[keys[0]][None] == suppl_table[keys[0]][:, None])
+    for key in keys[1:]:
+        intersect &= (table[key][None] == suppl_table[key][:, None])
+    i, j = np.argwhere(intersect).T
+
+    # entries not yet in supplement
+    append = (~np.in1d(table[keys[0]], suppl_table[keys[0]]))
+    for key in keys[1:]:
+        append |= (~np.in1d(table[key], suppl_table[key]))
+    k = np.argwhere(append).T
+
+    if not len(i) and not len(k[0]):
+        return
+
+    logger.info(f'updating "{path}" table in {filename}')
+
+    if len(i):
+        suppl_table[i] = table[j]
+    if len(k[0]):
+        suppl_table = vstack([suppl_table, table[k[0]]])
+
+    if not dry_run:
+        suppl_table.write(str(filename), format='hdf5', path=path, append=True, overwrite=True)
+        save_version(filename, path)
+    else:
+        logger.info('dry run, not saving anything')
+
+
+def get_table(filename, path, dtype=None, create=False):
+    """
+    Gets a table from an HDF5 file. Optionally creates the file if it does not exist.
```

Comment:
This comment is confusing because this function does not actually create a new file, instead it only issues the warning that the calling code will later create that file.

Reply:
Right. I edited the docstring, and also made this function private. I changed logging message, but still, this is a specific function, not general.